service/raft/group0_state_machine.cc

```diff
@@ -83,12 +123,98 @@ future<> group0_state_id_handler::advertise_state_id(utils::UUID state_id) {
+                                                return member_ip.has_value();
+                                            }) |
+                                            std::ranges::views::transform([this](const auto& member_ip) {
+                                                return _gc_times.find(member_ip.value());
+                                            }) |
+                                            std::ranges::views::filter([this](const auto& gc_time_it) {
+                                                return gc_time_it != _gc_times.end();
+                                            }) |
+                                            std::ranges::views::transform([](const auto& gc_time_it) {
+                                                return gc_time_it->second;
+                                            }) |
+                                            std::ranges::to<std::vector<gc_clock::time_point>>();
+
+            slogger.debug("Group0 members count: {}, filtered GC times count: {}", group0_members.size(), filtered_gc_times.size());
+
+            if (filtered_gc_times.size() < group0_members.size()) {
+                slogger.debug("Not all endpoints have reported their tombstone GC time yet");
+                return make_ready_future<>();
+            }
+
+            SCYLLA_ASSERT(!filtered_gc_times.empty());
+
+            const auto tombstone_gc_time = *std::ranges::min_element(filtered_gc_times,
+                                                                     [](const auto& a, const auto& b) {
+                                                                         return a < b;
+                                                                     })
+                                           // subtracting one resolution unit:
+                                           // - the GC time has 1 s resolution
+                                           // - there might be state_ids with the same GC time that didn't propagate yet
+                                           //   (and thus we don't want to GC these)
+                                           - gc_clock::duration{1};
+
+            if (!(_gc_time_last_reconcile < tombstone_gc_time)) {
+                slogger.debug("Skipping the current round due to the stale tombstone GC time: {}", tombstone_gc_time);
+                return make_ready_future();
+            }
+
+            _gc_time_last_reconcile = tombstone_gc_time;
+
+            slogger.debug("Setting reconcile time to {}", tombstone_gc_time);
+
+            auto& gc_state = _local_db.get().get_compaction_manager().get_tombstone_gc_state();
+            gc_state.update_group0_refresh_time(tombstone_gc_time);
+
+            return make_ready_future<>();
+        });
+}
+
+future<> group0_state_id_handler::on_remove(gms::inet_address endpoint, gms::permit_id pid) {
+    _gc_times.erase(endpoint);
```

Comment:
Regarding my previous comment to make `_gc_times` host ID based --

is we make it so, then in `on_remove` we might not be able to translate `endpoint` to host ID (I'm not sure, but we should consider the possibility)

but then consider my other comment about reacting to group 0 config changes -- perhaps this is the moment where we should remove an entry from `_gc_times` (i.e. when a node leaves group 0 config), not when it disappears from gossiper. Then it becomes a no-problem; group 0 config is already host ID based.

Reply:
yep, changed to react to the group0 configuration changes.